### PR TITLE
Fixed race condition in scan peripherals

### LINKED ISSUE #480: Sometimes startPeripheralScan doesn't return any result
Hi!
I am using your library for a personal project. The problem I have is that when I try to scan many times, the function startPeripheralScan doesn't return anything, or some devices aren't shown. Here is my code:
 
```
  void startScan() {
      _devicesList.clear();
      _devicesSink.add(null);
      _scanSubscription = _bleManager
         .startPeripheralScan(scanMode: ScanMode.balanced, allowDuplicates: false)
         .listen((device) {
             // exclude duplicated devices
             for (var d in _devicesList) {
                 if (device.peripheral.identifier == d.peripheral.identifier) return;
             }
             _devicesList.add(device);
            _devicesSink.add(_devicesList.sublist(0));
      });
      _timeoutTimer = new Timer(const Duration(seconds: 5), () => stopScan());
   }

   // Stop BLE devices scan
  Future<void> stopScan() async {
     _timeoutTimer?.cancel();
     if (_devicesList.length == 0) {
       _devicesSink.add([]);
     }
     await _scanSubscription?.cancel();
     await _bleManager?.stopPeripheralScan();
  }
```

### PR DESCRIPTION
**Pull request by dariuszseweryn**
Under certain circumstances a problem can be reported by the native side very quickly. Since monitoring for callbacks started after the call those responses could be delivered before anyone would start to listen.

Fixes #480

### CODE REVIEW

File: lib/src/bridge/scanning_mixin.dart [hunk 1]
```diff
@@ -1,42 +1,49 @@
...
-        ArgumentName.callbackType: callbackType,
-        ArgumentName.uuids: uuids,
-        ArgumentName.allowDuplicates: allowDuplicates,
-      },
-    );
-
+  Stream<ScanResult> startDeviceScan(int scanMode,
+      int callbackType,
+      List<String> uuids,
+      bool allowDuplicates,) {
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:24]: Run formatter, please

File: lib/src/bridge/scanning_mixin.dart [hunk 2]
```diff
@@ -1,42 +1,49 @@
...
+              MethodName.startDeviceScan,
+              <String, dynamic>{
+                ArgumentName.scanMode: scanMode,
+                ArgumentName.callbackType: callbackType,
+                ArgumentName.uuids: uuids,
+                ArgumentName.allowDuplicates: allowDuplicates,
+              },
+            ),
+        onCancel: () =>
+            _methodChannel.invokeMethod(MethodName.stopDeviceScan));
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:41]:
```suggestion
            stopDeviceScan();
```

File: lib/src/bridge/scanning_mixin.dart [hunk 3]
```diff
@@ -1,42 +1,49 @@
...
+
+    sc.addStream(_scanEvents)
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:43]:
```suggestion
    sc.addStream(_scanEvents, cancelOnError: true)
```

File: lib/src/bridge/scanning_mixin.dart (hunk 1 above)
Comment (dariuszseweryn) [reply lib/src/bridge/scanning_mixin.dart:24]: I did. Unless the formatter is not automatically loaded on the project?

File: lib/src/bridge/scanning_mixin.dart [hunk 4]
```diff
@@ -1,42 +1,47 @@
...
+        .receiveBroadcastStream()
+        .handleError(
+      (errorJson) {
+        throw BleError.fromJson(jsonDecode(errorJson.details));
+      },
+      test: (error) => error is PlatformException,
+    ).map((scanResultJson) => ScanResult.fromJson(
+              jsonDecode(scanResultJson),
+              _manager,
+            ));
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:17]:
```suggestion
            ),);
```

File: lib/src/bridge/scanning_mixin.dart [hunk 5]
```diff
@@ -1,42 +1,47 @@
...
+              <String, dynamic>{
+                ArgumentName.scanMode: scanMode,
+                ArgumentName.callbackType: callbackType,
+                ArgumentName.uuids: uuids,
+                ArgumentName.allowDuplicates: allowDuplicates,
+              },
+            ),
+        onCancel: () => stopDeviceScan());
+
+    sc.addStream(_scanEvents, cancelOnError: true).then((_) => sc?.close());
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:42]:
```suggestion
    sc.addStream(_scanEvents, cancelOnError: true,).then((_) => sc?.close(),);
```

File: lib/src/bridge/scanning_mixin.dart [hunk 6]
```diff
@@ -1,42 +1,48 @@
...
   void _prepareScanEventsStream() {
-    _scanEvents =
-        const EventChannel(ChannelName.scanningEvents).receiveBroadcastStream();
+    _scanEvents = const EventChannel(ChannelName.scanningEvents)
+        .receiveBroadcastStream()
+        .handleError(
+      (errorJson) {
+        throw BleError.fromJson(jsonDecode(errorJson.details));
+      },
+      test: (error) => error is PlatformException
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:13]:
```suggestion
      test: (error) => error is PlatformException,
```

File: lib/src/bridge/scanning_mixin.dart [hunk 7]
```diff
@@ -1,42 +1,48 @@
...
+          ArgumentName.scanMode: scanMode,
+          ArgumentName.callbackType: callbackType,
+          ArgumentName.uuids: uuids,
+          ArgumentName.allowDuplicates: allowDuplicates,
+        },
+      ),
+      onCancel: () => stopDeviceScan(),
+    );
+
+    sc.addStream(_scanEvents, cancelOnError: true).then((_) => sc?.close(),);
```
Comment (mikolak) [inline lib/src/bridge/scanning_mixin.dart:43]:
```suggestion
    sc.addStream(_scanEvents, cancelOnError: true,).then((_) => sc?.close(),);
```

**Pull request merged**